src/XAMLTools.Core/XAMLCombine/XAMLCombiner.cs

```diff
@@ -149,43 +161,99 @@ public string Combine(IEnumerable<string> sourceFiles, string targetFile)
 
-                        if (string.IsNullOrEmpty(key) == false)
-                        {
-                            // Check key unique
-                            if (keys.Contains(key))
+                            if (mergedDictionariesListNode == null)
                             {
                                 continue;
                             }
 
-                            keys.Add(key);
+                            var currentMergedSources = mergedDictionariesListNode.ChildNodes.OfType<XmlElement>()
+                                .Select(node => node.GetAttribute("Source"))
+                                .Where(source => !string.IsNullOrEmpty(source)).ToList();
+
+                            foreach (var mergedDictionaryReference in xmlElement.ChildNodes)
+                            {
+                                if (mergedDictionaryReference is not XmlElement mergedDictionaryReferenceElement || mergedDictionaryReferenceElement.Name != "ResourceDictionary")
+                                {
+                                    continue;
+                                }
+
+                                var sourceValue = mergedDictionaryReferenceElement.GetAttribute("Source");
+                                if (string.IsNullOrEmpty(sourceValue))
+                                {
+                                    continue;
+                                }
+
+                                // Check if it was already added
+                                if (currentMergedSources.Any(x => x.Equals(sourceValue)))
+                                {
+                                    continue;
+                                }
 
-                            // Create ResourceElement for key and XML node
-                            var res = new ResourceElement(key, importedElement, FillKeys(importedElement));
-                            resourceElements.Add(key, res);
-                            resourcesList.Add(res);
+                                // Import ResourceDictionary reference node from processed XML document to final XML document                        
+                                var importedResourceDictionaryReference = (XmlElement)finalDocument.ImportNode(mergedDictionaryReferenceElement, false);
+                                mergedDictionariesListNode.AppendChild(importedResourceDictionaryReference);
+                            }
+                        }
+
+                        continue;
+                    }
+
+                    // Resources
+
+                    // Import XML node from one XML document to result XML document                        
+                    var importedElement = (XmlElement)finalDocument.ImportNode(xmlElement, true);
+
+                    // Find resource key
+                    // TODO: Is any other variants???
+                    var key = string.Empty;
+                    if (importedElement.HasAttribute("Key"))
```

Comment:
I think we can replace with GetAttribute (should be safe based on reference code https://github.com/microsoft/referencesource/blob/master/System.Xml/System/Xml/Dom/XmlElement.cs#L226)